src/compiler/parser.ts

```diff
@@ -3087,14 +3087,15 @@ namespace ts {
         }
 
         function parseUnionOrIntersectionType(kind: SyntaxKind.UnionType | SyntaxKind.IntersectionType, parseConstituentType: () => TypeNode, operator: SyntaxKind.BarToken | SyntaxKind.AmpersandToken): TypeNode {
-            parseOptional(operator);
+            const leadingToken = parseOptionalToken(operator);
             let type = parseConstituentType();
+            const startPos = leadingToken ? leadingToken.pos : type.pos;

```

Comment:
Can you add a test for this output. It seems like this change would result in missing comments:

```ts
type A = 
  /*leading0*/
  | /*leading1*/ 1 /*trailing1*/ 
  | /*leading2*/ 2 /*trailing2*/;
```

When we emit this declaration today, we include the leading and trailing comments for each type in the union, but do not emit `/*leading0*/`. I would like to see the test verify that we still preserve the comments around each type and whether this will now also include the leading comment before the first `|`.

Reply:
Also, I would move this line inside of the `if` statement that follows it? There is no need to determine `startPos` if we end up not using it.